UI/system-info-posix.cpp

```diff
@@ -1,6 +1,672 @@
 #include "system-info.hpp"
+#include "util/dstr.hpp"
+#include "util/platform.h"
+#include <sys/utsname.h>
+#include <unistd.h>
+#include <string>
+#include <fstream>
+#include <regex>
+
+using namespace std;
+
+extern "C" {
+#include "pci/pci.h"
+}
+
+// Anonymous namespace ensures internal linkage
+namespace {
+struct drm_card_info {
+	uint16_t vendor_id;
+	uint16_t device_id;
+	uint16_t subsystem_vendor;
+	uint16_t subsystem_device;
+
+	/* match_count is the number of matches between
+	 * the tokenized GPU identification string and
+	 * the device_name and vendor_name supplied by
+	 * libpci. It is used to associate the GPU that
+	 * OBS is using and the cards found in a bus scan
+	 * using libpci.
+	 */
+	uint16_t match_count;
+
+	/* The following 2 fields are easily
+	 * obtained for AMD GPUs. Reporting
+	 * for NVIDIA GPUs is not working at
+	 * the moment.
+	 */
+	uint64_t dedicated_vram_total;
+	uint64_t shared_system_memory_total;
+
+	// PCI domain:bus:slot:function
+	std::string dbsf;
+	std::string device_name;
+	std::string vendor_name;
+};
+
+constexpr std::string_view WHITE_SPACE = " \f\n\r\t\v";
+
+// trim_ws() will remove leading and trailing
+// white space from the string.
+void trim_ws(std::string &s)
+{
+	// Trim leading whitespace
+	size_t pos = s.find_first_not_of(WHITE_SPACE);
+	if (pos != std::string::npos)
+		s = s.substr(pos);
+
+	// Trim trailing whitespace
+	pos = s.find_last_not_of(WHITE_SPACE);
+	if (pos != std::string::npos)
+		s = s.substr(0, pos + 1);
+}
+
+bool compare_match_strength(const drm_card_info &a, const drm_card_info &b)
+{
+	return a.match_count > b.match_count;
+}
+
+void adjust_gpu_model(std::string &model)
+{
+	/* Use the sub-string between the [] brackets. For example,
+	 * the NVIDIA Quadro P4000 model string from PCI ID database
+	 * is "GP104GL [Quadro P4000]", and we only want the "Quadro
+	 * P4000" sub-string.
+	 */
+	size_t first = model.find('[');
+	size_t last = model.find_last_of(']');
+	if ((last - first) > 1) {
+		std::string adjusted_model = model.substr(first + 1, last - first - 1);
+		model.assign(adjusted_model);
+	}
+}
+
+bool get_distribution_info(std::string &distro, std::string &version)
+{
+	ifstream file;
+	std::string line;
+	const std::string systemd_file = "/etc/os-release";
+	const std::string flatpak_file = "/.flatpak-info";
+
+	distro = "";
+	version = "";
+
+	if (std::filesystem::exists(systemd_file)) {
```

Comment:
Wrong priority, Flatpak should be checked first.

Some runtime add an os-release file for compatibility, but has been proven to be an issue (e.g. 6.2 was still showing 5.15 in os-release).

Reply:
```suggestion
	if (std::filesystem::exists(flatpak_file)) {
		/* The .flatpak-info file has a line of the form:
		 *
		 * runtime=runtime/org.kde.Platform/x86_64/6.6
		 *
		 * Parse the line into tokens to identify the name and
		 * version, which are "org.kde.Platform" and "6.6" respectively,
		 * in the example below.
		 */
		file.open(flatpak_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 16, "runtime=runtime/") == 0) {
					size_t pos = line.find('/');
					if (pos != string::npos && line.at(pos + 1) != '\0') {
						line.erase(0, pos + 1);

						/* Split the string into tokens with a regex
						 * a regex of one or more '/' characters'.
						 */
						std::regex fp_reg("[/]+");
						vector<std::string> fp_tokens(
							sregex_token_iterator(line.begin(), line.end(), fp_reg, -1),
							sregex_token_iterator());
						if (fp_tokens.size() >= 2) {
							auto token = begin(fp_tokens);
							distro = "Flatpak " + *token;
							token = next(fp_tokens.end(), -1);
							version = *token;
						} else {
							distro = "Flatpak unknown";
							version = "0";
							blog(LOG_DEBUG,
							     "%s: Format of 'runtime' entry unrecognized in file %s",
							     __FUNCTION__, flatpak_file.c_str());
						}
						break;
					}
				}
			}
			file.close();
		}
	} else if (std::filesystem::exists(systemd_file)) {
		/* systemd-based distributions use /etc/os-release to identify
		 * the OS. For example, the Ubuntu 24.04.1 variant looks like:
		 *
		 * PRETTY_NAME="Ubuntu 24.04.1 LTS"
		 * NAME="Ubuntu"
		 * VERSION_ID="24.04"
		 * VERSION="24.04.1 LTS (Noble Numbat)"
		 * VERSION_CODENAME=noble
		 * ID=ubuntu
		 * ID_LIKE=debian
		 * HOME_URL="https://www.ubuntu.com/"
		 * SUPPORT_URL="https://help.ubuntu.com/"
		 * BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
		 * PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
		 * UBUNTU_CODENAME=noble
		 * LOGO=ubuntu-logo
		 *
		 * Parse the file looking for the NAME and VERSION_ID fields.
		 * Note that some distributions wrap the entry in '"' characters
		 * while others do not, so we need to remove those characters.
		 */
		file.open(systemd_file);
		if (file.is_open()) {
			while (getline(file, line)) {
				if (line.compare(0, 4, "NAME") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						distro = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						distro.erase(std::remove(distro.begin(), distro.end(), '"'),
							     distro.end());

						trim_ws(distro);
						continue;
					}
				}

				if (line.compare(0, 10, "VERSION_ID") == 0) {
					size_t pos = line.find('=');
					if (pos != std::string::npos && line.at(pos + 1) != '\0') {
						version = line.substr(pos + 1);

						// Remove the '"' characters from the string, if any.
						version.erase(std::remove(version.begin(), version.end(), '"'),
							      version.end());

						trim_ws(version);
						continue;
					}
				}
			}
			file.close();
		}
	} else {
		blog(LOG_DEBUG, "%s: Failed to find host OS or flatpak info", __FUNCTION__);
		return false;
	}

	return true;
```